multicluster/controllers/multicluster/memberclusterannounce_controller.go

```diff
@@ -256,30 +257,40 @@ func (r *MemberClusterAnnounceReconciler) AddMember(MemberId common.ClusterID) {
 		Reason:             ReasonNeverConnected,
 	})
 
-	r.memberStatus[MemberId] = &multiclusterv1alpha1.ClusterStatus{ClusterID: string(MemberId),
+	r.memberStatus[memberID] = &multiclusterv1alpha1.ClusterStatus{ClusterID: string(memberID),
 		Conditions: conditions}
 
-	r.timerData[MemberId] = &timerData{connected: false, lastUpdateTime: time.Time{}}
+	r.timerData[memberID] = &timerData{connected: false, lastUpdateTime: time.Time{}}
+	klog.InfoS("new member is added", "member", memberID)
 }
 
-func (r *MemberClusterAnnounceReconciler) RemoveMember(MemberId common.ClusterID) {
-	defer r.mapLock.Unlock()
+func (r *MemberClusterAnnounceReconciler) RemoveMember(memberID common.ClusterID) {
 	r.mapLock.Lock()
+	defer r.mapLock.Unlock()
 
-	if _, ok := r.memberStatus[MemberId]; ok {
-		delete(r.memberStatus, MemberId)
+	if _, ok := r.memberStatus[memberID]; ok {
+		delete(r.memberStatus, memberID)
 		return
 	}
 
-	if _, ok := r.timerData[MemberId]; ok {
-		delete(r.timerData, MemberId)
+	if _, ok := r.timerData[memberID]; ok {
+		klog.InfoS("remove members", "member", memberID)
+		delete(r.timerData, memberID)
```

Comment:
I think the previous code doesn't have data race. The issue was caused by data inconsistency and it's still there.
`RemoveMember` removed the item from `r.memberStatus` but not `r.timerData`, but `processMCSStatus` expects items existing in `r.timerData` also exist in `r.memberStatus`, so accessing `r.memberStatus[member].Condition` panic.

I think the right fix is removing the two `return` in this function. We should add unit test in this PR, not a separate PR. One of the points of tests is to expose the situation that has the issue, and to prove this patch can fix it. I don't an e2e test can be very helpful here, it's likely unable to reproduce the scenario.

Reply:
Fixed, thanks.